File: backend/src/patrones_de_diseño/factory/ruby/shapefact.rb
#!/bin/ruby

class Shape
    def factory(type)
        if (type == "Circle"); return Circle.new; end;
        if (type == "Square"); return Square.new; end;
        puts "Bad shape creation: %s" % type
        raise NotImplementedError
    end

    # Método para obtener las clases hijas
    def self.descendants
        ObjectSpace.each_object(Class).select { |klass| klass < self }
    end
end

class Circle < Shape
    def draw; puts ("Circle.draw"); end;
    def erase; puts ("Circle.erase"); end;
end

class Square < Shape
	def draw; puts ("Square.draw"); end;
    def erase; puts ("Square.erase"); end;
end

def shapeNameGen(n)
	types = Shape.descendants()
	response = []

	for i in (1..n) do
		response.push(types.sample.to_s)
	end
	return response
end

S = Shape.new

for shape in shapeNameGen(7) do
	shape = S.factory(shape)
    shape.draw
    shape.erase
end

=begin
Fuentes:
http://www.eq8.eu/blogs/13-ruby-ancestors-descendants-and-other-annoying-relatives
=end

File: backend/src/patrones_de_diseño/factory/python/shapefact.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Fuente:
# http://python-3-patterns-idioms-test.readthedocs.io/en/latest/Factory.html

import random

class Shape(object):
    # Creación basada en el nombre de la clase:
    @staticmethod
    def factory(shape_type):
        try:
            return eval(shape_type + "()")
        except NameError as e:
            print("Bad shape creation: %s" % shape_type)
            print(e)

        #if type == "Circle": return Circle()
        #if type == "Square": return Square()
        #raise Exception("Bad shape creation: " + type)

class Circle(Shape):
    def draw(self): print("Circle.draw")
    def erase(self): print("Circle.erase")

class Square(Shape):
    def draw(self): print("Square.draw")
    def erase(self): print("Square.erase")

# Genera los nombres de clase al azar:
def shapeNameGen(n):
    types = Shape.__subclasses__()
            #[<class '__main__.Circle'>, 
            # <class '__main__.Square'>]
    for i in range(n):
        yield random.choice(types).__name__

shapes = \
  [ Shape.factory(i) for i in shapeNameGen(7)]

for shape in shapes:
    shape.draw()
    shape.erase()